File: code/av12453_poly.py
#!/usr/bin/env python3
"""Polynomial-time enumeration of Av(12453), hence its Wilf class.

The recurrence can be derived by using 12453 = 12 direct-sum 231 and
reading a permutation from left to right.  Every completed increasing pair
creates a requirement that a certain future upper subsequence avoid 231.
The unfinished recursive
intervals of these 231-avoidance requirements form a composition-valued
stack L.  This is the recurrence published as the Biers-Ariel program linked
from OEIS A116485.

This program compresses that composition-valued state.  Write H(i,j,L),
where L is either the empty composition or a tuple of positive parts.  The
key quantity below is

    K[i,j,l][u,v],

the number of recursion paths which start with control (i,j) and top stack
part l, process everything before the marked old tail (including material
later fused into the head), and expose that unchanged tail with control
(u,v).

All dependencies have rank one less, where rank = j + (sum of stack parts),
so the kernels can be filled in increasing order of j+l.  If max_n=N, the
implementation stores the convenient padded region j+l <= N+2 (the tight
cutoff is N+1).  The final empty-stack values G(i,j)
are then filled in increasing j, and |Av_n(12453)| = G(n+1,n+2).

The implementation deliberately mirrors the proof recurrence rather than
trying to optimize constants.  It uses polynomially many exact-integer
operations (a coarse bound is O(N^8) arithmetic operations and O(N^5)
integer storage).
"""

from __future__ import annotations

import argparse
import time
from collections import defaultdict
from itertools import combinations, permutations
from typing import DefaultDict, Dict, Iterable, Mapping, Tuple

Control = Tuple[int, int]
Row = Dict[Control, int]
KernelKey = Tuple[int, int, int]


def _add_scaled(target: DefaultDict[Control, int], row: Mapping[Control, int],
                scale: int = 1) -> None:
    """Add ``scale * row`` to a sparse row vector."""
    for state, value in row.items():
        target[state] += scale * value


def count_avoiders(max_n: int) -> Tuple[list[int], int, int]:
    """Return |Av_n(12453)| for 0 <= n <= max_n.

    The two extra returned integers are the number of kernel rows and the
    number of stored nonzero kernel entries.  They are useful for checking
    the polynomial state compression empirically.
    """
    if max_n < 0:
        raise ValueError("max_n must be nonnegative")

    limit = max_n + 2
    kernels: Dict[KernelKey, Row] = {}

    # K_l((i,j),-), in increasing rank j+l.  Every right-hand-side kernel
    # has smaller rank.  Controls always satisfy 1 <= i < j.
    for rank in range(3, limit + 1):
        for j in range(2, rank):
            ell = rank - j
            for i in range(1, j):
                row: DefaultDict[Control, int] = defaultdict(int)

                # H(i,j,L) -> H(k,j-1,L), 1 <= k < i.
                for k in range(1, i):
                    _add_scaled(row, kernels[k, j - 1, ell])

                # H(i,j,lL) -> H(i,k,(l+j-k-1)L), i < k < j.
                for k in range(i + 1, j):
                    delta = j - k - 1
                    _add_scaled(row, kernels[i, k, ell + delta])

                # Decrement the top part.  At ell=1 this exposes the old
                # tail immediately, without changing the control state.
                if ell == 1:
                    row[i, j] += 1
                else:
                    _add_scaled(row, kernels[i, j, ell - 1], 2)

                # Split ell into a,b with a+b=ell-1.  The transfer for a is
                # followed by the transfer for b, i.e. sparse row-by-matrix
                # multiplication.
                for a in range(1, ell - 1):
                    b = ell - 1 - a
                    for middle, multiplicity in kernels[i, j, a].items():
                        _add_scaled(
                            row,
                            kernels[middle[0], middle[1], b],
                            multiplicity,
                        )

                kernels[i, j, ell] = dict(row)

    # Empty-stack values.  The unique terminal state is (1,2).  All other
    # dependencies have strictly smaller second coordinate, so increasing j
    # is a topological order.
    empty: Dict[Control, int] = {}
    for j in range(2, limit + 1):
        for i in range(1, j):
            if (i, j) == (1, 2):
                empty[i, j] = 1
                continue

            value = 0
            for k in range(1, i):
                value += empty[k, j - 1]

            for k in range(i + 1, j):
                delta = j - k - 1
                if delta == 0:
                    value += empty[i, k]
                else:
                    value += sum(
                        multiplicity * empty[end]
                        for end, multiplicity in kernels[i, k, delta].items()
                    )

            empty[i, j] = value

    coefficients = [empty[n + 1, n + 2] for n in range(max_n + 1)]
    nonzero_entries = sum(len(row) for row in kernels.values())
    return coefficients, len(kernels), nonzero_entries


def _avoids_12453(permutation: Tuple[int, ...]) -> bool:
    """Direct pattern test, used only by the optional small-n verifier."""
    pattern = (1, 2, 4, 5, 3)
    for indices in combinations(range(len(permutation)), 5):
        values = tuple(permutation[index] for index in indices)
        reduction = tuple(
            1 + sum(other < value for other in values) for value in values
        )
        if reduction == pattern:
            return False
    return True


def verify_by_brute_force(coefficients: list[int], through_n: int) -> None:
    """Raise AssertionError unless the requested small terms pass brute force."""
    through_n = min(through_n, len(coefficients) - 1)
    for n in range(through_n + 1):
        direct = sum(_avoids_12453(p) for p in permutations(range(1, n + 1)))
        if direct != coefficients[n]:
            raise AssertionError(
                f"n={n}: transfer algorithm gave {coefficients[n]}, "
                f"brute force gave {direct}"
            )


def main(argv: Iterable[str] | None = None) -> None:
    parser = argparse.ArgumentParser(
        description="Compute the Av(12453)=Av(31245) counting sequence."
    )
    parser.add_argument("max_n", nargs="?", type=int, default=20)
    parser.add_argument(
        "--verify-small",
        type=int,
        default=-1,
        metavar="N",
        help="also check terms through N by direct permutation enumeration",
    )
    args = parser.parse_args(argv)

    started = time.perf_counter()
    coefficients, rows, entries = count_avoiders(args.max_n)
    elapsed = time.perf_counter() - started
    if args.verify_small >= 0:
        verify_by_brute_force(coefficients, args.verify_small)
    for n, value in enumerate(coefficients):
        print(f"{n} {value}")
    print(
        f"# kernel_rows={rows} nonzero_entries={entries} "
        f"seconds={elapsed:.3f}"
    )
    if args.verify_small >= 0:
        checked = min(args.verify_small, args.max_n)
        print(f"# brute_force_verified_through={checked}")


if __name__ == "__main__":
    main()

File: code/av12453_poly.cpp
// Faster C++ implementation of the transfer-kernel recurrence described in
// av12453_polytime.tex.  Compile with:
//   g++ -O3 -std=c++20 av12453_poly.cpp -o av12453_poly

#include <algorithm>
#include <chrono>
#include <cstdint>
#include <cstdlib>
#include <iostream>
#include <utility>
#include <vector>

// A tiny nonnegative big integer is enough here and keeps this verifier
// dependency-free.  Coefficients are stored in base 10^9.
class Big {
  public:
    static constexpr std::uint32_t base = 1000000000U;
    std::vector<std::uint32_t> digits;

    Big(std::uint64_t value = 0) {
        while (value) {
            digits.push_back(static_cast<std::uint32_t>(value % base));
            value /= base;
        }
    }

    bool zero() const { return digits.empty(); }

    Big& operator+=(const Big& other) {
        const std::size_t size = std::max(digits.size(), other.digits.size());
        digits.resize(size, 0);
        std::uint64_t carry = 0;
        for (std::size_t p = 0; p < size; ++p) {
            const std::uint64_t sum = static_cast<std::uint64_t>(digits[p])
                + (p < other.digits.size() ? other.digits[p] : 0) + carry;
            digits[p] = static_cast<std::uint32_t>(sum % base);
            carry = sum / base;
        }
        if (carry) digits.push_back(static_cast<std::uint32_t>(carry));
        return *this;
    }

    friend Big operator+(Big left, const Big& right) {
        left += right;
        return left;
    }

    friend Big operator*(const Big& left, const Big& right) {
        if (left.zero() || right.zero()) return Big();
        Big product;
        product.digits.assign(left.digits.size() + right.digits.size(), 0);
        for (std::size_t a = 0; a < left.digits.size(); ++a) {
            std::uint64_t carry = 0;
            for (std::size_t b = 0; b < right.digits.size(); ++b) {
                const std::size_t p = a + b;
                const std::uint64_t value =
                    static_cast<std::uint64_t>(left.digits[a]) * right.digits[b]
                    + product.digits[p] + carry;
                product.digits[p] = static_cast<std::uint32_t>(value % base);
                carry = value / base;
            }
            std::size_t p = a + right.digits.size();
            while (carry) {
                if (p == product.digits.size()) product.digits.push_back(0);
                const std::uint64_t value = product.digits[p] + carry;
                product.digits[p] = static_cast<std::uint32_t>(value % base);
                carry = value / base;
                ++p;
            }
        }
        while (!product.digits.empty() && product.digits.back() == 0)
            product.digits.pop_back();
        return product;
    }

    friend bool operator==(const Big& value, int small) {
        return small == 0 && value.zero();
    }

    friend std::ostream& operator<<(std::ostream& out, const Big& value) {
        if (value.zero()) return out << '0';
        out << value.digits.back();
        for (std::size_t p = value.digits.size() - 1; p-- > 0;) {
            const std::uint32_t block = value.digits[p];
            std::uint32_t divisor = 100000000U;
            while (divisor > block && divisor > 1) {
                out << '0';
                divisor /= 10;
            }
            out << block;
        }
        return out;
    }
};
using Entry = std::pair<int, Big>;
using Row = std::vector<Entry>;

static int control_id(int i, int j) {
    return (j - 1) * (j - 2) / 2 + (i - 1);
}

int main(int argc, char** argv) {
    const int max_n = argc > 1 ? std::atoi(argv[1]) : 30;
    if (max_n < 0) return 2;
    const int limit = max_n + 2;
    const int controls = limit * (limit - 1) / 2;

    std::vector<int> id_i(controls), id_j(controls);
    for (int j = 2; j <= limit; ++j) {
        for (int i = 1; i < j; ++i) {
            const int id = control_id(i, j);
            id_i[id] = i;
            id_j[id] = j;
        }
    }

    // kernel[i][j][ell] is a sparse row indexed by endpoint control IDs.
    std::vector<std::vector<std::vector<Row>>> kernel(
        limit + 1,
        std::vector<std::vector<Row>>(
            limit + 1, std::vector<Row>(limit + 1)));

    std::vector<Big> accumulator(controls);
    std::vector<unsigned char> marked(controls, 0);
    std::vector<int> touched;
    touched.reserve(controls);

    auto touch_add = [&](int endpoint, const Big& value) {
        if (value == 0) return;
        if (!marked[endpoint]) {
            marked[endpoint] = 1;
            touched.push_back(endpoint);
        }
        accumulator[endpoint] += value;
    };

    auto add_row = [&](const Row& source) {
        for (const auto& [endpoint, value] : source) touch_add(endpoint, value);
    };

    auto add_row_twice = [&](const Row& source) {
        for (const auto& [endpoint, value] : source)
            touch_add(endpoint, value + value);
    };

    auto add_scaled_row = [&](const Row& source, const Big& scale) {
        for (const auto& [endpoint, value] : source)
            touch_add(endpoint, scale * value);
    };

    const auto started = std::chrono::steady_clock::now();
    std::size_t nonzero_entries = 0;
    std::size_t kernel_rows = 0;

    for (int rank = 3; rank <= limit; ++rank) {
        for (int j = 2; j < rank; ++j) {
            const int ell = rank - j;
            for (int i = 1; i < j; ++i) {
                touched.clear();

                for (int k = 1; k < i; ++k)
                    add_row(kernel[k][j - 1][ell]);

                for (int k = i + 1; k < j; ++k) {
                    const int delta = j - k - 1;
                    add_row(kernel[i][k][ell + delta]);
                }

                if (ell == 1) {
                    touch_add(control_id(i, j), Big(1));
                } else {
                    add_row_twice(kernel[i][j][ell - 1]);
                }

                for (int a = 1; a < ell - 1; ++a) {
                    const int b = ell - 1 - a;
                    for (const auto& [middle, multiplicity] :
                         kernel[i][j][a]) {
                        add_scaled_row(
                            kernel[id_i[middle]][id_j[middle]][b],
                            multiplicity);
                    }
                }

                std::sort(touched.begin(), touched.end());
                Row& output = kernel[i][j][ell];
                output.reserve(touched.size());
                for (int endpoint : touched) {
                    if (accumulator[endpoint] != 0)
                        output.emplace_back(endpoint,
                                            std::move(accumulator[endpoint]));
                    accumulator[endpoint] = 0;
                    marked[endpoint] = 0;
                }
                ++kernel_rows;
                nonzero_entries += output.size();
            }
        }
    }

    std::vector<Big> empty(controls);
    for (int j = 2; j <= limit; ++j) {
        for (int i = 1; i < j; ++i) {
            const int source = control_id(i, j);
            if (i == 1 && j == 2) {
                empty[source] = 1;
                continue;
            }
            Big value = 0;
            for (int k = 1; k < i; ++k)
                value += empty[control_id(k, j - 1)];
            for (int k = i + 1; k < j; ++k) {
                const int delta = j - k - 1;
                if (delta == 0) {
                    value += empty[control_id(i, k)];
                } else {
                    for (const auto& [endpoint, multiplicity] :
                         kernel[i][k][delta])
                        value += multiplicity * empty[endpoint];
                }
            }
            empty[source] = std::move(value);
        }
    }

    for (int n = 0; n <= max_n; ++n)
        std::cout << n << ' ' << empty[control_id(n + 1, n + 2)] << '\n';

    const auto finished = std::chrono::steady_clock::now();
    const double seconds =
        std::chrono::duration<double>(finished - started).count();
    std::cout << "# kernel_rows=" << kernel_rows
              << " nonzero_entries=" << nonzero_entries
              << " seconds=" << seconds << '\n';
}
